pos_loyalty_exclude/models/loyalty_rule.py

```diff
@@ -0,0 +1,31 @@
+# Copyright 2024 Camptocamp (https://www.camptocamp.com).
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, models
+from odoo.osv import expression
+
+
+class LoyaltyRule(models.Model):
+    _inherit = "loyalty.rule"
+
+    @api.depends("product_ids", "product_category_id", "product_tag_id")
+    def _compute_valid_product_ids(self):
+        super()._compute_valid_product_ids()
+        # exclude product when no config in loyalty rule
+        for rule in self:
+            if not (
+                rule.product_ids
+                or rule.product_category_id
+                or rule.product_tag_id
+                or rule.product_domain not in ("[]", "[['sale_ok', '=', True]]")
+            ):
+                rule.any_product = False
+                rule.valid_product_ids = self.env["product.product"].search(
```

Comment:
if the reason you do a new search here instead of filtering already defined `rule.valid_product_ids` is a concern for performance, then a comment would be welcome

Reply:
Hi anh, If I understand correctly, when the rule has configuration on product/categories..., the reward applies on all products. However, looks like a bug as it returns an empty array, because of this https://github.com/odoo/odoo/blob/35305419c53dc19e995c7dada8d51c94cb93dc68/addons/pos_loyalty/models/loyalty_rule.py#L40

Reply:
ah ok got it actually looks like a smart optimization: `any_product` to avoid loading all products in `valid_product_ids`